diesel/src/mysql/connection/result.rs

```diff
@@ -0,0 +1,37 @@
+extern crate mysqlclient_sys as ffi;
+
+pub struct RawResult(*mut ffi::MYSQL_RES);
+
+impl RawResult {
+    pub unsafe fn from_raw(raw: *mut ffi::MYSQL_RES) -> Option<Self> {
+        if raw.is_null() {
+            None
+        } else {
+            Some(RawResult(raw))
+        }
+    }
+
+    pub fn fields(&mut self) -> ResultFields {
+        unsafe { ffi::mysql_field_seek(self.0, 0); }
+        ResultFields(self)
+    }
+}
+
+impl Drop for RawResult {
+    fn drop(&mut self) {
+        unsafe { ffi::mysql_free_result(self.0); }
+    }
+}
+
+pub struct ResultFields<'a>(&'a mut RawResult);
+
+impl<'a> Iterator for ResultFields<'a> {
+    type Item = &'a ffi::MYSQL_FIELD;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            ffi::mysql_fetch_field((self.0).0)
+                .as_ref()
```

Comment:
The lifetime is the `'a` referenced in the `Self::Item` type, which is the lifetime of the `RawResult` reference we are holding onto, which is the actual correct lifetime. The fields are fully disjoint, so we don't need any sort of streaming iterator abstraction here.